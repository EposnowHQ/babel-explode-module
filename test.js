// @flow
'use strict';
const t = require('babel-types');
const explodeModule = require('./');
const pluginTester = require('babel-plugin-tester');
const printAST = require('ast-pretty-print');
const getBabylonOptions = require('babylon-options');

const babelOptions = {
  parserOpts: getBabylonOptions({
    stage: 0,
    plugins: ['flow'],
  }),
};

const plugin = () => {
  return {
    name: 'babel-explode-module',
    visitor: {
      Program(path) {
        let exploded = explodeModule(path.node);
        let printed = printAST(exploded) + '\n';

        path.replaceWith(
          t.program([
            t.expressionStatement(
              t.templateLiteral([
                t.templateElement({ raw: printed }, true)
              ], [])
            )
          ])
        );
        path.stop();
      },
    },
  };
};

pluginTester({
  title: 'export',
  plugin,
  snapshot: true,
  babelOptions,
  tests: {
    'default binding': {code: 'export default a;'},
    'default object': {code: 'export default {a};'},
    'default array': {code: 'export default [a];'},
    'default number': {code: 'export default 1;'},
    'default string': {code: 'export default "a";'},
    'default boolean': {code: 'export default true;'},
    'default expression': {code: 'export default a ? b : c;'},
    'default named function': {code: 'export default function a() {}'},
    'default function': {code: 'export default function() {}'},
    'default named class': {code: 'export default class a {}'},
    'default class': {code: 'export default class {}'},
    'named specifier': {code: 'export {a}'},
    'named specifier multiple': {code: 'export {a, b}'},
    'named specifier renamed': {code: 'export {a as b}'},
    'named function': {code: 'export function a() {}'},
    'named class': {code: 'export class a {}'},
    'named var': {code: 'export var a;'},
    'named var multiple': {code: 'export var a, b;'},
    'named let': {code: 'export let a;'},
    'from default': {code: 'export default from "b";'},
    'from default renamed': {code: 'export a from "b";'},
    'from all': {code: 'export * from "b";'},
    'from named': {code: 'export {a} from "b";'},
    'from named multiple': {code: 'export {a, b} from "b";'},
    'from named renamed': {code: 'export {a as b} from "c";'},
    'from named default renamed': {code: 'export {default as a} from "b";'},
  },
});

pluginTester({
  title: 'imports',
  plugin,
  snapshot: true,
  babelOptions,
  tests: {
    default: {code: 'import a from "b";'},
    named: {code: 'import {a} from "b";'},
    'named renamed': {code: 'import {a as b} from "c";'},
    namespace: {code: 'import * as a from "b";'},
    'default and named': {code: 'import a, {b} from "c";'},
    effect: {code: 'import "c";'},
    'import type': {code: 'import type a from "b";'},
    'import typeof': {code: 'import typeof a from "b";'},
    'import type inner': {code: 'import {type a} from "b";'},
    'import typeof inner': {code: 'import {typeof a} from "b";'},
    'import value/type/typeof inner': {
      code: 'import {a, type b, typeof c} from "b";',
    },
  },
});

pluginTester({
  title: 'statements',
  plugin,
  snapshot: true,
  babelOptions,
  tests: {
    empty: {code: ';'},
    variable: {code: 'var a;'},
    function: {code: 'function a() {}'},
    class: {code: 'class a {}'},
    'variable multiple': {code: 'var a, b;'},
    for: {code: 'for (;;) {}'},
    while: {code: 'while (a) {}'},
    'type alias': {code: 'type a = {};'},
    interface: {code: 'interface a {}'},
  },
});
