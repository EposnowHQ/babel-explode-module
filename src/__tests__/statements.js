// @flow
import {testSuite} from '../../test/helpers';

testSuite({
  name: 'statements',
  tests: {
    empty: {code: ';'},
    variable: {code: 'var a;'},
    function: {code: 'function a() {}'},
    class: {code: 'class a {}'},
    'variable multiple': {code: 'var a, b;'},
    for: {code: 'for (;;) {}'},
    while: {code: 'while (a) {}'},
    'type alias': {code: 'type a = {};'},
    interface: {code: 'interface a {}'},
  },
});
