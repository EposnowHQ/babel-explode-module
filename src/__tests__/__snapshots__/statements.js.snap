// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`statements class 1`] = `
"
code:

class a {}

exploded:

  exports: []
  imports: []
  statements:
    - ClassDeclaration (1:0,1:10)
        body: ClassBody (1:8,1:10)
          body: []
        id: Identifier (1:6,1:7)
          name: 'a'
        superClass: null"
`;

exports[`statements empty 1`] = `
"
code:

;

exploded:

  exports: []
  imports: []
  statements: []"
`;

exports[`statements for 1`] = `
"
code:

for (;;) {}

exploded:

  exports: []
  imports: []
  statements:
    - ForStatement (1:0,1:11)
        body: BlockStatement (1:9,1:11)
          body: []
          directives: []
        init: null
        test: null
        update: null"
`;

exports[`statements function 1`] = `
"
code:

function a() {}

exploded:

  exports: []
  imports: []
  statements:
    - FunctionDeclaration (1:0,1:15)
        async: false
        body: BlockStatement (1:13,1:15)
          body: []
          directives: []
        expression: false
        generator: false
        id: Identifier (1:9,1:10)
          name: 'a'
        params: []"
`;

exports[`statements interface 1`] = `
"
code:

interface a {}

exploded:

  exports: []
  imports: []
  statements:
    - InterfaceDeclaration (1:0,1:14)
        body: ObjectTypeAnnotation (1:12,1:14)
          callProperties: []
          exact: false
          indexers: []
          properties: []
        extends: []
        id: Identifier (1:10,1:11)
          name: 'a'
        mixins: []
        typeParameters: null"
`;

exports[`statements type alias 1`] = `
"
code:

type a = {};

exploded:

  exports: []
  imports: []
  statements:
    - TypeAlias (1:0,1:12)
        id: Identifier (1:5,1:6)
          name: 'a'
        right: ObjectTypeAnnotation (1:9,1:11)
          callProperties: []
          exact: false
          indexers: []
          properties: []
        typeParameters: null"
`;

exports[`statements variable 1`] = `
"
code:

var a;

exploded:

  exports: []
  imports: []
  statements:
    - VariableDeclaration (1:4,1:5)
        declarations:
          - VariableDeclarator (1:4,1:5)
              id: Identifier (1:4,1:5)
                name: 'a'
              init: null
        innerComments: undefined
        kind: 'var'
        leadingComments: undefined
        trailingComments: undefined"
`;

exports[`statements variable multiple 1`] = `
"
code:

var a, b;

exploded:

  exports: []
  imports: []
  statements:
    - VariableDeclaration (1:4,1:5)
        declarations:
          - VariableDeclarator (1:4,1:5)
              id: Identifier (1:4,1:5)
                name: 'a'
              init: null
        innerComments: undefined
        kind: 'var'
        leadingComments: undefined
        trailingComments: undefined
    - VariableDeclaration (1:7,1:8)
        declarations:
          - VariableDeclarator (1:7,1:8)
              id: Identifier (1:7,1:8)
                name: 'b'
              init: null
        innerComments: undefined
        kind: 'var'
        leadingComments: undefined
        trailingComments: undefined"
`;

exports[`statements while 1`] = `
"
code:

while (a) {}

exploded:

  exports: []
  imports: []
  statements:
    - WhileStatement (1:0,1:12)
        body: BlockStatement (1:10,1:12)
          body: []
          directives: []
        test: Identifier (1:7,1:8)
          name: 'a'"
`;
